sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -69,6 +69,7 @@ class Analyzer(catalog: Catalog,
       typeCoercionRules ++
       extendedRules : _*),
     Batch("Check Analysis", Once,
+      CheckCast ::
```

Comment:
If we check resolution first, 
1 check cast will be invalid since legitimate cast expression must be not resolved.
2 ```select cast(key as binary) from src``` will give a ```scala.NotImplementedError```

how about not change the order but catch exception of child.dataType to make it pass this rule?